File: Baekjoon/Dynamic Programming/9251.py
n, m = input(), input()
w, h = len(n), len(m)
dp = [[0 for _ in range(w+1)] for _ in range(h+1)]

for i in range(1, h+1):
    for j in range(1, w+1):
        if m[i-1] == n[j-1]:
            dp[i][j] = dp[i-1][j-1] + 1
        else:
            dp[i][j] = max(dp[i-1][j], dp[i][j-1])
        
print(dp[-1][-1])

File: Baekjoon/Dynamic Programming/9251.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.max

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine()
    val m = br.readLine()
    val w = n.length
    val h = m.length
    val dp = Array(h+1){ Array(w+1){0} }

    for(i in 1..h) {
        for (j in 1.. w) {
            if (m[i-1] == n[j-1]) {
                dp[i][j] = dp[i-1][j-1] + 1
            }
            else {
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])
            }
        }
    }

    println(dp[h][w])
}